pkg/server/service.go

```diff
@@ -129,11 +134,13 @@ func NewCRIContainerdService(config criconfig.Config, client *containerd.Client)
 	c.imageFSPath = imageFSPath(config.ContainerdRootDir, config.ContainerdConfig.Snapshotter)
 	logrus.Infof("Get image filesystem path %q", c.imageFSPath)
 
-	c.netPlugin, err = ocicni.InitCNI(config.NetworkPluginConfDir, config.NetworkPluginBinDir)
-	if err != nil {
-		return nil, fmt.Errorf("failed to initialize cni plugin: %v", err)
+	c.netPlugin, err = cni.New(cni.WithMinNetworkCount(2),
+		cni.WithPluginConfDir(config.NetworkPluginConfDir),
+		cni.WithPluginDir([]string{config.NetworkPluginBinDir}),
+		cni.WithLoNetwork())
+	if err != nil && !cni.IsCNINotInitialized(err) {
```

Comment:
I'm not very comfortable with leaving `netPlugin` as `nil` if CNI is not initialized.
1) There will be race condition as I mentioned in https://github.com/containerd/cri/pull/663/files#r174359472.
2) If functions other than `Status` is called, this may cause a panic. Especially kubelet assume that host network pod could run even when network is not ready, and this assume that `netPlugin` will not be accessed at all if pod is using host network.
3) Even if someone try to run normal pod when network plugin is not ready, containerd should not panic.

Anyway, I feel like it is better to have netPlugin here, but add another function to return whether the netPlugin is ready or not.

Reply:
In what scenario does the kublet would use netplugin for host network ? We can definitely add a check if thats the case. 
The reason we stuck to New is for better error handling instead of ignoring the errors. I am not sure about the race. We can discuss about this.

Reply:
CRI is a grpc service, all functions could be called at anytime as long as client sends a request.
There are several functions using `netPlugin`. We need a lock to protect it if we want to use it in one function, and set it in another function.

We don't necessarily need to ignore the error. We just need to define error better. .To me, missing CNI plugins/config is not an error, it is a state of cni library. We can just return the `netPlugin` in `UnInitialized` state. As for other errors, of course, we can return error.

Reply:
I think the issue with the way it is. We return the cni struct and error. essentially
```
cni:=CNI{......}
err:= filnetworkConfig(....)
if err != nil{
    return cni,err
}
return cni,nil

Reply:
This is what it will end up looking. If we need to lock it , we can lock it. I was of the opinion same as yours. But after some discussion with @crosbymichael I feel that in order to make it work in cri , we end up return an uninitialized state as above ^.

Reply:
My argument is that CRI should not impose bad API design for libs that it consumes.  A lib's `New` should either return a fully initialized struct or error.  Not something inbetween that its incapable of doing work.  It just adds complexity and bugs.  Before, if there are invalid configs,etc. all the load errors are discarded.  How is this ever good in any universe?